### PR TITLE
DP-442 BE- Share my information - Answering Declaration form (Forms API)

### PR DESCRIPTION
**Pull request by glgoaco**
[DP-442](https://noticingsystem.atlassian.net/browse/DP-442)

Extending the Forms API to support new question types and tightly coupling SharedConsent with AnswerSets.

### CODE REVIEW

File: Services/CO.CDP.Forms.WebApi/UseCase/UpdateFormSectionAnswersUseCase.cs [hunk 1]
```diff
@@ -30,26 +30,47 @@ public async Task<bool> Execute((Guid formId, Guid sectionId, Guid answerSetId,
 
         var existingAnswerSet = await formRepository.GetFormAnswerSetAsync(sectionId, organisationId, answerSetId);
 
-        await UploadFileIfRequired(answers, questionDictionary, existingAnswerSet);
```
Comment (shilpigoel-goaco) [inline Services/CO.CDP.Forms.WebApi/UseCase/UpdateFormSectionAnswersUseCase.cs:33]: why uploading file is removed?

**Review by shilpigoel-goaco — APPROVED** [review]

File: Services/CO.CDP.OrganisationInformation.Persistence/Forms/FormAnswer.cs [hunk 2]
```diff
@@ -55,11 +58,7 @@ public class FormAnswerSet : IEntityDate
 {
     public int Id { get; set; }
     public required Guid Guid { get; set; }
-
-    [ForeignKey(nameof(Organisation))]
-    public required int OrganisationId { get; set; }
-    public required Organisation Organisation { get; set; }
-
+    public required SharedConsent SharedConsent { get; set; }
```
Comment (jakzal) [inline Services/CO.CDP.OrganisationInformation.Persistence/Forms/FormAnswer.cs:61]: Do we need it here? Just a thought. Assuming we'll be always querying for FormAnswers and AnswerSets by SharedConsent, defining the relationship on one side might be sufficient?

**Comment by jakzal** [comment]
Will the migrations apply cleanly if there's anything in the form_answer_sets table?

File: Services/CO.CDP.OrganisationInformation.Persistence/Forms/FormAnswer.cs (hunk 2 above)
Comment (glgoaco) [reply Services/CO.CDP.OrganisationInformation.Persistence/Forms/FormAnswer.cs:61]: I think its needed

File: Services/CO.CDP.Forms.WebApi/UseCase/UpdateFormSectionAnswersUseCase.cs [hunk 3]
```diff
@@ -38,18 +38,41 @@ public async Task<bool> Execute((Guid formId, Guid sectionId, Guid answerSetId,
...
-                Guid = answerSetId,
+                throw new Exception($"Form not found.");
+            }
+            if (section.Form.Guid != formId)
+            {
+                throw new Exception($"Form with Guid {formId} not found.");
+            }
+
+
+            var sharedConsent = new Persistence.SharedConsent
```
Comment (dharmverma) [inline Services/CO.CDP.Forms.WebApi/UseCase/UpdateFormSectionAnswersUseCase.cs:51]: This doesn't seems right, we should not be creating SharedConsent every time a new AnswerSet is created.
It should only be created  when no draft SharedConset exists

File: Services/CO.CDP.Forms.WebApi/UseCase/UpdateFormSectionAnswersUseCase.cs [hunk 4]
```diff
@@ -38,18 +38,41 @@ public async Task<bool> Execute((Guid formId, Guid sectionId, Guid answerSetId,
...
+            {
+                Guid = Guid.NewGuid(),
                 OrganisationId = organisation.Id,
                 Organisation = organisation,
+                Form = section.Form,
+                SubmissionState = Persistence.SubmissionState.Draft,
+                SubmittedAt = DateTimeOffset.UtcNow,
```
Comment (dharmverma) [inline Services/CO.CDP.Forms.WebApi/UseCase/UpdateFormSectionAnswersUseCase.cs:58]: Do we need additional SubmittedAt column, Can''t we use CreatedOn column?

File: Services/CO.CDP.Forms.WebApi/UseCase/UpdateFormSectionAnswersUseCase.cs [hunk 5]
```diff
@@ -38,18 +38,41 @@ public async Task<bool> Execute((Guid formId, Guid sectionId, Guid answerSetId,
...
+                FormVersionId = String.Empty,
+                BookingReference = string.Empty,
```
Comment (dharmverma) [inline Services/CO.CDP.Forms.WebApi/UseCase/UpdateFormSectionAnswersUseCase.cs:60]: Do we need to explicitly set BookingReference to Empty?

File: Services/CO.CDP.Forms.WebApi/UseCase/UpdateFormSectionAnswersUseCase.cs [hunk 6]
```diff
@@ -38,18 +38,41 @@ public async Task<bool> Execute((Guid formId, Guid sectionId, Guid answerSetId,
...
+                AnswerSets = new List<Persistence.FormAnswerSet>()
+            };
+
+            existingAnswerSet = new Persistence.FormAnswerSet
+            {
+                Guid = answerSetId,
+                SharedConsent = sharedConsent,
                 Section = section,
-                Answers = MapAnswers(answers, questionDictionary, []),
+                Answers = MapAnswers(answers, questionDictionary, new List<Persistence.FormAnswer>()),
```
[the remark this answers is not in the document]
Comment (glgoaco) [reply Services/CO.CDP.Forms.WebApi/UseCase/UpdateFormSectionAnswersUseCase.cs:69]: lets go for consistency over best practise

File: Services/CO.CDP.Forms.WebApi/UseCase/UpdateFormSectionAnswersUseCase.cs (hunk 3 above)
Comment (glgoaco) [reply Services/CO.CDP.Forms.WebApi/UseCase/UpdateFormSectionAnswersUseCase.cs:51]: added an explicit check as discussed in the thread

File: Services/CO.CDP.Forms.WebApi/UseCase/UpdateFormSectionAnswersUseCase.cs [hunk 7]
```diff
@@ -38,11 +38,22 @@ public async Task<bool> Execute((Guid formId, Guid sectionId, Guid answerSetId,
         }
```
Comment (dharmverma) [inline Services/CO.CDP.Forms.WebApi/UseCase/UpdateFormSectionAnswersUseCase.cs:38]: We also need to duplicate the AnswerSets/Answer, rather update when SharedConsent booking reference is already generated (not in draft mode)

File: Services/CO.CDP.OrganisationInformation.Persistence/DatabaseFormRepository.cs
Comment (dharmverma) [inline Services/CO.CDP.OrganisationInformation.Persistence/DatabaseFormRepository.cs:1]: Avoid using Region to group methods please
https://softwareengineering.stackexchange.com/questions/53086/are-regions-an-antipattern-or-code-smell

File: Services/CO.CDP.Forms.WebApi/UseCase/UpdateFormSectionAnswersUseCase.cs (hunk 7 above)
Comment (glgoaco) [reply Services/CO.CDP.Forms.WebApi/UseCase/UpdateFormSectionAnswersUseCase.cs:38]: This is already discussed and not part of Dp-442

File: Services/CO.CDP.OrganisationInformation.Persistence/DatabaseFormRepository.cs
Comment (glgoaco) [reply Services/CO.CDP.OrganisationInformation.Persistence/DatabaseFormRepository.cs:1]: This wssnt added in this PR

File: Services/CO.CDP.Forms.WebApi/UseCase/UpdateFormSectionAnswersUseCase.cs (hunk 7 above)
Comment (dharmverma) [reply Services/CO.CDP.Forms.WebApi/UseCase/UpdateFormSectionAnswersUseCase.cs:38]: Have we got a ticket for that? please add it to the backlog if not there, and add what was discussed.

**Review by dbgoaco — APPROVED** [review]

File: Services/CO.CDP.Forms.WebApi/UseCase/UpdateFormSectionAnswersUseCase.cs [hunk 8]
```diff
@@ -23,32 +23,19 @@ public async Task<bool> Execute((Guid formId, Guid sectionId, Guid answerSetId,
...
-                Guid = answerSetId,
-                OrganisationId = organisation.Id,
-                Organisation = organisation,
-                Section = section,
-                Answers = MapAnswers(answers, questionDictionary, []),
-            };
-        }
-
-        await formRepository.SaveAnswerSet(existingAnswerSet);
+        await UploadFileIfRequired(answers, questionDictionary, answerSet);
```
Comment (dharmverma) [inline Services/CO.CDP.Forms.WebApi/UseCase/UpdateFormSectionAnswersUseCase.cs:37]: UploadFileIfRequired will not work now, it should be done before Answers are mapped.

**Review by dharmverma — APPROVED** [review]

File: Services/CO.CDP.Forms.WebApi/UseCase/UpdateFormSectionAnswersUseCase.cs (hunk 8 above)
Comment (jakzal) [reply Services/CO.CDP.Forms.WebApi/UseCase/UpdateFormSectionAnswersUseCase.cs:36]: Updated, thanks!

**Pull request merged**